synapse/handlers/message.py

```diff
@@ -965,6 +969,13 @@ async def create_new_client_event(
         if builder.internal_metadata.outlier:
             event.internal_metadata.outlier = True
             context = EventContext.for_outlier()
+        elif (
+            event.type == EventTypes.MSC2716_INSERTION
+            and full_state_ids_at_event
+            and builder.internal_metadata.is_historical()
+        ):
+            old_state = await self.store.get_events_as_list(full_state_ids_at_event)
```

Comment:
@MadLittleMods, @erikjohnston: I'm trying to figure out how this works, so that I can fit it in with my work on faster joins and partial room state.

so... how does this work? `full_state_ids_at_event` is set to `auth_event_ids`, which is absolutely *not* the full state at the event. Rather, as the docstring says, it is "The event ids to use as the auth_events for the new event." - ie, the subset of the state that is needed to auth the event.

Reply:
The relevant call stack for this is:

```
RoomBatchSendEventRestServlet.on_POST
handle_batch_of_events
persist_historical_events
create_event
create_new_client_event
```

In `RoomBatchSendEventRestServlet.on_POST` we [fetch the full state for the event](https://github.com/matrix-org/synapse/blob/c56bfb08bc071368db23f3b1c593724eb4f205f0/synapse/rest/client/room_batch.py#L126-L132) we're inserting next to (specified by `/batch_send?prev_event_id=xxx`) as a base to start from. Then as we persist the `state_events_at_start`, we also add that to the `auth_event_ids` list.

That full `auth_event_ids` list is then passed down that call stack and gets stripped down to only include the necessary bits in the code here. And we use `full_state_ids_at_event` to "Add explicit state to the insertion event so the rest of the batch can inherit the same state and `state_group`"

---

Here is the context where we originally started stripping down the state: https://github.com/matrix-org/synapse/pull/9247#discussion_r651803946

Reply:
right, thanks for the explanation, but this seems extremely confusing. At the very least, if `create_new_client_event` and co are expecting `auth_event_ids` to be the full state rather than just the auth events, then the docstrings should be updated to say so - right now they are flat-out wrong.

But generally, this feels like an abuse of `auth_event_ids` - we shouldn't be using it to mean "normally the auth events, except under special circumstances, when it is the full state". I'm not sure I have any immediate suggestions, but adding a new parameter would be better than the current situation.

@erikjohnston I think you've been following the MSC2716 work a bit more than me - any thoughts here?

Reply:
Ah, I think I got confused when reading this as at some point the state at the insertion event was limited to the used auth events, which is no longer the case. 

I agree that we should therefore be passing the full state through as a separate parameter. Though at some point I'm wondering if this code path needs refactoring to make it easier to use in the `/send_batch` case, as its starting to get a bit unwieldy with all the different cases.

Reply:
@richvdh Do you want me to tackle this new full state parameter? Don't want to collide and cause conflicts with your existing refactor.

Reply:
You're unlikely to conflict with me here, so please do go ahead!